sycl/include/CL/sycl/kernel_handler.hpp

```diff
@@ -0,0 +1,72 @@
+//==------ kernel_handler.hpp -- SYCL standard header file -----*- C++ -*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// This guard is needed because the libsycl.so can compiled with C++ <=14
+// while the code requires C++17. This code is not supposed to be used by the
+// libsycl.so so it should not be a problem.
+#if __cplusplus > 201402L
+template <auto &S> struct specialization_id_name_generator {};
+#endif
+
+} // namespace detail
+
+/// Reading the value of a specialization constant
+///
+/// \ingroup sycl_api
+class kernel_handler {
+public:
+#if __cplusplus > 201402L
+  template <auto &S>
+  typename std::remove_reference_t<decltype(S)> get_specialization_constant() {
+#ifdef __SYCL_DEVICE_ONLY__
+    return getSpecializationConstantOnDevice<S>();
+#else
+    throw cl::sycl::feature_not_supported(
+        "kernel_handler::get_specialization_constant() is not supported by "
+        "host device.",
+        PI_INVALID_OPERATION);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+#endif // __cplusplus > 201402L
+
+  void __init_specialization_constants_buffer(
```

Comment:
Can we make `__init*` function private?

Tagging @elizabethandrews

Reply:
Based on the current spec, the kernel_handler object generated by the compiler calls this function. If you make this private, I'm not sure how object accesses it. 

Having said that I see that the `__init` function for `stream`, etc is declared `private `in headers. I do not know how that works considering compiler generated kernel object clone calls these. So I may be misunderstanding something. Tagging @erichkeane.

Reply:
I think the way we look it up in SemaSYCL we don't end up considering access specifiers, we just do a lookup, so `private` doesn't mean anything in that case.

Reply:
Thanks, in this case moved to `private`.